syntax = "proto3";
package todo;

import "google/protobuf/timestamp.proto";

service TodoService {
    rpc AddTodo(Todo) returns (TodoResponse) {}
    rpc GetTodo(Todo) returns (TodoResponse) {}
    rpc ListAllTodo(EmptyRequest) returns (stream ListAllTodoResponse) {}
    rpc DeleteTodo(Todo) returns (TodoResponse) {}
    rpc EditTodo(Todo) returns (TodoResponse) {}
    rpc ToggleStatus(Todo) returns (TodoResponse) {}
}

message EmptyRequest {}

message ListAllTodoResponse {
    OperationErrors status = 1;
    Todo todo = 2;
}

message TodoResponse {
    OperationErrors status = 1;
    Todo todo = 2;
}

message Todo {
    int32 id = 1;
    string title = 2;
    string description = 3;
    TodoStatus status = 4;
    google.protobuf.Timestamp created_at = 5;
    google.protobuf.Timestamp updated_at = 6;
}

message OperationErrors {
    Status code = 1;
    repeated string errors = 2;
}

enum Status {
    OK = 0;
    ERROR = 1;
    NOT_FOUND = 2;
    UPDATED = 4;
    DELETED = 5;
}

enum TodoStatus {
    NEW = 0;
    PROGRESS = 1;
    COMPLETED = 2;
}
